### PR TITLE
Introduce EJBCA UpstreamAuthority plugin for SPIRE Server

### PR DESCRIPTION
**Pull request by m8rmclaren**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**

This PR introduces [EJBCA](https://www.ejbca.org/) as a Server UpstreamAuthority plugin. This plugin uses a connected EJBCA to issue intermediate certificates for the SPIRE server.

> Per #4163, the EJBCA UpstreamAuthority plugin is compatible with [EJBCA Community](https://www.ejbca.org/).

**Description of change**
* Adds EJBCA as a built-in UpstreamAuthority plugin.
* Adds the [EJBCA Go Client SDK](https://github.com/Keyfactor/ejbca-go-client-sdk/tree/main) as a required dependency for communicating with the EJBCA REST API.

> Original PR was #5201 which was closed to rename the base branch from `main` to a more descriptive name.

**Which issue this PR fixes**

### CODE REVIEW

File: doc/plugin_server_upstreamauthority_ejbca.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+| `end_entity_name`          | (optional) The name of the end entity, or configuration for how the EJBCA UpstreamAuthority should determine the end entity name. See [End Entity Name Customization](#ejbca-end-entity-name-customization-leaf-certificates) for more info. |                                    |
+| `account_binding_id`       | (optional) An account binding ID in EJBCA to associate with issued certificates.                                                                                                                                                             |                                    |
+
+> Configuration parameters that have an override from Environment Variables will always override the provided value from the SPIRE configuration with the values in the environment.
+>
+> If all configuration parameters for the selected auth method are specified by environment variables, an empty block still must exist to select the auth method.
+
+```hcl
+UpstreamAuthority "ejbca" {
```
Comment (amartinezfayo) [inline doc/plugin_server_upstreamauthority_ejbca.md:35]: We need to update conf/server/server_full.conf to include a sample configuration.

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca.go [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+	}
+
+	logger.Debug("Preparing EJBCA enrollment request")
+	password, err := generateRandomString(16)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to generate random password: %v", err)
+	}
+	enrollConfig := ejbcaclient.EnrollCertificateRestRequest{}
+	enrollConfig.SetUsername(endEntityName)
+	enrollConfig.SetPassword(password)
```
Comment (amartinezfayo) [inline pkg/server/plugin/upstreamauthority/ejbca/ejbca.go:164]: How is this password conveyed with the server?

**Review by amartinezfayo — COMMENTED** [review]
Thank you @m8rmclaren for addressing the previous comments. I've added some more comments/suggestions.

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca.go [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+			if block == nil {
+				return status.Error(codes.Internal, "failed to parse CA certificate PEM")
+			}
+			caBytes = append(caBytes, block.Bytes...)
+		}
+	case enrollResponse.GetResponseFormat() == "DER":
+		logger.Debug("EJBCA returned certificate in DER format - serializing")
+
+		bytes := []byte(enrollResponse.GetCertificate())
+		bytes, err := base64.StdEncoding.DecodeString(string(bytes))
```
[the remark this answers is not in the document]
Comment (m8rmclaren) [reply pkg/server/plugin/upstreamauthority/ejbca/ejbca.go:210]: Nice catch, thanks

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca.go [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+	if err != nil {
+		return status.Errorf(codes.Internal, "unable to determine end entity name: %v", err)
+	}
+
+	logger.Debug("Preparing EJBCA enrollment request")
+	password, err := generateRandomString(16)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to generate random password: %v", err)
+	}
+	enrollConfig := ejbcaclient.EnrollCertificateRestRequest{}
```
[the remark this answers is not in the document]
Comment (m8rmclaren) [reply pkg/server/plugin/upstreamauthority/ejbca/ejbca.go:162]: It absolutely could.

File: doc/plugin_server_upstreamauthority_ejbca.md (hunk 1 above)
Comment (m8rmclaren) [reply doc/plugin_server_upstreamauthority_ejbca.md:35]: I missed this - thanks

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca_client.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	if config.ClientCertPath == "" {
+		config.ClientCertPath = p.hooks.getEnv("EJBCA_CLIENT_CERT_PATH")
+	}
+	if config.ClientKeyPath == "" {
+		config.ClientKeyPath = p.hooks.getEnv("EJBCA_CLIENT_CERT_KEY_PATH")
+	}
+
+	if config.ClientCertPath == "" {
```
[the remark this answers is not in the document]
Comment (m8rmclaren) [reply pkg/server/plugin/upstreamauthority/ejbca/ejbca_client.go:36]: Hi @amartinezfayo!
I wrote it in this order to set the corresponding variable from the environment if it wasn't set in the conf file, and if neither is defined, return an error.

If there is a more elegant way to set `ClientCertPath` and `ClientCertKeyPath` from the environment if not present in the conf file, I'm open to suggestions.

For now, I just wrote more descriptive comments.

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca.go (hunk 2 above)
Comment (m8rmclaren) [reply pkg/server/plugin/upstreamauthority/ejbca/ejbca.go:164]: Hi @amartinezfayo!

This password (enrollment code) is conveyed to EJBCA in the `POST /v1/certificate/pkcs10enroll` request and has no direct effect concerning the EJBCA plugin.

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca.go [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+	if len(caChain) == 0 {
+		return status.Error(codes.Internal, "EJBCA did not return a CA chain")
+	}
+
+	rootCa := caChain[len(caChain)-1]
+	logger.Debug("Retrieved root CA from CA chain", "rootCa", rootCa.Subject.String(), "intermediates", len(caChain)-1)
+
+	// x509CertificateChain contains the leaf CA certificate, then any intermediates up to but not including the root CA.
+	x509CertificateAuthorityChain, err := x509certificate.ToPluginProtos(append([]*x509.Certificate{cert}, caChain[:len(caChain)-1]...))
```
Comment (amartinezfayo) [inline pkg/server/plugin/upstreamauthority/ejbca/ejbca.go:244]:
```suggestion
	x509CertificateAuthorityChain, err := x509certificate.ToPluginFromCertificates(append([]*x509.Certificate{cert}, caChain[:len(caChain)-1]...))
```

File: pkg/server/plugin/upstreamauthority/ejbca/ejbca.go [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to serialize certificate chain: %v", err)
+	}
+
+	rootCACertificate, err := x509certificate.ToPluginProtos([]*x509.Certificate{rootCa})
```
Comment (amartinezfayo) [inline pkg/server/plugin/upstreamauthority/ejbca/ejbca.go:249]:
```suggestion
	rootCACertificate, err := x509certificate.ToPluginFromCertificates([]*x509.Certificate{rootCa})
```

**Review by amartinezfayo — COMMENTED** [review]
Thank you @m8rmclaren for your patience and addressing the comments.
He have introduced some recent changes (#5340) that requires replacing the use of x509certificate.ToPluginProtos with x509certificate.ToPluginFromCertificates.
I think that we should be good to go after updating that.

**Comment by m8rmclaren** [comment]
Hi @amartinezfayo!

Thank you for pointing this out; I've updated all usages of `ToPluginProtos` to `ToPluginFromCertificates`. The tests are now passing in my [GitHub Actions](https://github.com/m8rmclaren/spire/actions/runs/10686696488).

**Review by amartinezfayo — APPROVED** [review]
Thank you @m8rmclaren for this contribution!

**Pull request merged**